rust/rbac-registration/src/cardano/cip509/rbac/metadata.rs

```diff
@@ -0,0 +1,230 @@
+//! Cip509 RBAC metadata.
+
+use std::collections::HashMap;
+
+use catalyst_types::problem_report::ProblemReport;
+use minicbor::{decode, Decode, Decoder};
+use strum_macros::FromRepr;
+
+use crate::{
+    cardano::cip509::{
+        rbac::{
+            role_data::RoleNumberAndData, C509Cert, RoleData, RoleNumber, SimplePublicKeyType,
+            X509DerCert,
+        },
+        types::CertKeyHash,
+        utils::Cip0134UriSet,
+    },
+    utils::decode_helper::{
+        decode_any, decode_array_len, decode_bytes, decode_helper, decode_map_len,
+        report_duplicated_key,
+    },
+};
+
+/// Cip509 RBAC metadata.
+///
+/// See [this document] for more details.
+///
+/// [this document]: https://github.com/input-output-hk/catalyst-CIPs/tree/x509-role-registration-metadata/CIP-XXXX
+#[derive(Debug, PartialEq, Clone)]
+#[allow(clippy::module_name_repetitions)]
+pub struct Cip509RbacMetadata {
+    /// A potentially empty list of x509 certificates.
+    pub x509_certs: Vec<X509DerCert>,
+    /// A potentially empty list of c509 certificates.
+    pub c509_certs: Vec<C509Cert>,
+    /// A set of URIs contained in both x509 and c509 certificates.
+    ///
+    /// URIs from different certificate types are stored separately and certificate
+    /// indexes are preserved too.
+    ///
+    /// This field isn't present in the encoded format and is populated by processing both
+    /// `x509_certs` and `c509_certs` fields.
+    pub certificate_uris: Cip0134UriSet,
+    /// A list of public keys that can be used instead of storing full certificates.
+    ///
+    /// Check [this section] to understand the how certificates and the public keys list
+    /// are related.
+    ///
+    /// [this section]: https://github.com/input-output-hk/catalyst-CIPs/tree/x509-role-registration-metadata/CIP-XXXX#storing-certificates-and-public-key
+    pub pub_keys: Vec<SimplePublicKeyType>,
+    /// A potentially empty list of revoked certificates.
+    pub revocation_list: Vec<CertKeyHash>,
+    /// A potentially empty role data.
+    pub role_data: HashMap<RoleNumber, RoleData>,
```

Comment:
We already store the role data as a map in the registration chain, so I think here it would be better to revert my changes and have a structure that follows CBOR format just to simplify decoding unless we want to have a map here too.